### PR TITLE
KMS Writeback Fixes

### LINKED ISSUE #4874: KMS (DRM) writeback device stalls, test suite fails (on Pi4, at least)
### Describe the bug

The vc4 GPU driver includes a "virtual" writeback connector which is great but it doesn't seem to work; attempted frame updates with writeback always time out.

### Steps to reproduce the behaviour

(Bug report evolved from this conversation: https://forums.raspberrypi.com/viewtopic.php?t=328068)

The easiest repro is using the [kms_writeback tests in igt-gpu-tools](https://gitlab.freedesktop.org/drm/igt-gpu-tools/-/blob/master/tests/kms_writeback.c). On my system, at least, this fails every time (though other KMS/DRM tests pass, so the test suite in general does work). It seems the vblank event simply never arrives, and various error messages show up in the dmesg.

### Device (s)

Raspberry Pi 4 Mod. B

### System

```
% raspinfo                                                        ~ egnor-pi4-5
System Information
------------------

Raspberry Pi 4 Model B Rev 1.4
PRETTY_NAME="Raspbian GNU/Linux 11 (bullseye)"
NAME="Raspbian GNU/Linux"
VERSION_ID="11"
VERSION="11 (bullseye)"

Raspberry Pi reference 2021-05-07
Generated using pi-gen, https://github.com/RPi-Distro/pi-gen, dcfd74d7d1fa293065ac6d565711e9ff891fe2b8, stage4

Linux egnor-pi4-5 5.10.63-v7l+ #1496 SMP Wed Dec 1 15:58:56 GMT 2021 armv7l GNU/Linux
Revision	: d03114
Serial		: 100000003365054b
Model		: Raspberry Pi 4 Model B Rev 1.4
Throttled flag  : throttled=0x0
Camera          : supported=1 detected=0

Videocore information
---------------------

Dec  1 2021 15:02:25 
Copyright (c) 2012 Broadcom
version 71bd3109023a0c8575585ba87cbb374d2eeb038f (clean) (release) (start_db)

alloc failures:     0
compactions:        0
legacy block fails: 0

Filesystem information
----------------------
Filesystem     1K-blocks     Used Available Use% Mounted on
/dev/root      120332468 15743232  98453636  14% /
devtmpfs         3780788        0   3780788   0% /dev
tmpfs            4043956        0   4043956   0% /dev/shm
tmpfs            1617584      992   1616592   1% /run
tmpfs               5120        4      5116   1% /run/lock
/dev/mmcblk0p6    258094    49437    208658  20% /boot
tmpfs             808788       28    808760   1% /run/user/1000

Filename				Type		Size		Used	Priority
/var/swap                               file		102396		1536	-2

Package version information
---------------------------
raspberrypi-ui-mods:
  Installed: 1.20211123
raspberrypi-sys-mods:
  Installed: 20220110+1
openbox:
  Installed: 3.6.1-9+rpt1+deb11u1
lxpanel:
  Installed: 0.10.1-2+rpt11
pcmanfm:
  Installed: 1.3.2-1+rpt6
rpd-plym-splash:
  Installed: 0.29

Networking Information
----------------------

/usr/bin/raspinfo: line 67: ifconfig: command not found

USB Information
---------------

/:  Bus 02.Port 1: Dev 1, Class=root_hub, Driver=xhci_hcd/4p, 5000M
/:  Bus 01.Port 1: Dev 1, Class=root_hub, Driver=xhci_hcd/1p, 480M
    |__ Port 1: Dev 2, If 0, Class=Hub, Driver=hub/4p, 480M
        |__ Port 3: Dev 3, If 0, Class=Human Interface Device, Driver=usbhid, 12M
        |__ Port 3: Dev 3, If 1, Class=Human Interface Device, Driver=usbhid, 12M
        |__ Port 3: Dev 3, If 2, Class=Human Interface Device, Driver=usbhid, 12M

config.txt
----------

arm_freq=1500
audio_pwm_mode=514
camera_auto_detect=-1
config_hdmi_boost=5
core_freq=500
core_freq_min=200
disable_commandline_tags=2
disable_l2cache=1
disable_overscan=1
disable_splash=1
display_hdmi_rotate=-1
display_lcd_rotate=-1
dvfs=3
enable_gic=1
force_eeprom_read=1
force_pwm_open=1
framebuffer_ignore_alpha=1
framebuffer_swap=1
gpu_freq=500
gpu_freq_min=250
init_uart_clock=0x2dc6c00
lcd_framerate=60
mask_gpu_interrupt0=3072
mask_gpu_interrupt1=25635
max_framebuffers=2
over_voltage=2
over_voltage_avs=-10000
pause_burst_frames=1
program_serial_random=1
second_boot=1
total_mem=8192
hdmi_force_cec_address:0=65535
hdmi_force_cec_address:1=65535
hdmi_pixel_freq_limit:0=0x11e1a300
hdmi_pixel_freq_limit:1=0x11e1a300
device_tree=-
overlay_prefix=overlays/
hdmi_cvt:0=
hdmi_cvt:1=
hdmi_edid_filename:0=
hdmi_edid_filename:1=
hdmi_timings:0=
hdmi_timings:1=

cmdline.txt
-----------
coherent_pool=1M 8250.nr_uarts=0 snd_bcm2835.enable_compat_alsa=0 snd_bcm2835.enable_hdmi=1 video=HDMI-A-1:1920x1080M@60 smsc95xx.macaddr=DC:A6:32:EE:13:D7 vc_mem.mem_base=0x3ec00000 vc_mem.mem_size=0x40000000  console=ttyS0,115200 console=tty1 root=/dev/mmcblk0p7 rootfstype=ext4 elevator=deadline fsck.repair=yes rootwait quiet splash plymouth.ignore-serial-consoles

raspi-gpio settings
-------------------

BANK0 (GPIO 0 to 27):
GPIO 0: level=1 fsel=0 func=INPUT pull=UP
GPIO 1: level=1 fsel=0 func=INPUT pull=UP
GPIO 2: level=1 fsel=0 func=INPUT pull=UP
GPIO 3: level=1 fsel=0 func=INPUT pull=UP
GPIO 4: level=1 fsel=0 func=INPUT pull=UP
GPIO 5: level=1 fsel=0 func=INPUT pull=UP
GPIO 6: level=1 fsel=0 func=INPUT pull=UP
GPIO 7: level=1 fsel=0 func=INPUT pull=UP
GPIO 8: level=1 fsel=0 func=INPUT pull=UP
GPIO 9: level=0 fsel=0 func=INPUT pull=DOWN
GPIO 10: level=0 fsel=0 func=INPUT pull=DOWN
GPIO 11: level=0 fsel=0 func=INPUT pull=DOWN
GPIO 12: level=0 fsel=0 func=INPUT pull=DOWN
GPIO 13: level=0 fsel=0 func=INPUT pull=DOWN
GPIO 14: level=1 fsel=0 func=INPUT pull=NONE
GPIO 15: level=1 fsel=0 func=INPUT pull=UP
GPIO 16: level=0 fsel=0 func=INPUT pull=DOWN
GPIO 17: level=0 fsel=0 func=INPUT pull=DOWN
GPIO 18: level=0 fsel=0 func=INPUT pull=DOWN
GPIO 19: level=0 fsel=0 func=INPUT pull=DOWN
GPIO 20: level=0 fsel=0 func=INPUT pull=DOWN
GPIO 21: level=0 fsel=0 func=INPUT pull=DOWN
GPIO 22: level=0 fsel=0 func=INPUT pull=DOWN
GPIO 23: level=0 fsel=0 func=INPUT pull=DOWN
GPIO 24: level=0 fsel=0 func=INPUT pull=DOWN
GPIO 25: level=0 fsel=0 func=INPUT pull=DOWN
GPIO 26: level=0 fsel=0 func=INPUT pull=DOWN
GPIO 27: level=0 fsel=0 func=INPUT pull=DOWN
BANK1 (GPIO 28 to 45):
GPIO 28: level=1 fsel=2 alt=5 func=RGMII_MDIO pull=UP
GPIO 29: level=0 fsel=2 alt=5 func=RGMII_MDC pull=DOWN
GPIO 30: level=0 fsel=7 alt=3 func=CTS0 pull=UP
GPIO 31: level=0 fsel=7 alt=3 func=RTS0 pull=NONE
GPIO 32: level=1 fsel=7 alt=3 func=TXD0 pull=NONE
GPIO 33: level=1 fsel=7 alt=3 func=RXD0 pull=UP
GPIO 34: level=0 fsel=7 alt=3 func=SD1_CLK pull=NONE
GPIO 35: level=1 fsel=7 alt=3 func=SD1_CMD pull=UP
GPIO 36: level=1 fsel=7 alt=3 func=SD1_DAT0 pull=UP
GPIO 37: level=1 fsel=7 alt=3 func=SD1_DAT1 pull=UP
GPIO 38: level=1 fsel=7 alt=3 func=SD1_DAT2 pull=UP
GPIO 39: level=1 fsel=7 alt=3 func=SD1_DAT3 pull=UP
GPIO 40: level=0 fsel=4 alt=0 func=PWM1_0 pull=NONE
GPIO 41: level=1 fsel=4 alt=0 func=PWM1_1 pull=NONE
GPIO 42: level=0 fsel=1 func=OUTPUT pull=UP
GPIO 43: level=1 fsel=0 func=INPUT pull=UP
GPIO 44: level=1 fsel=0 func=INPUT pull=UP
GPIO 45: level=1 fsel=0 func=INPUT pull=UP
BANK2 (GPIO 46 to 53):
GPIO 46: level=0 fsel=0 func=INPUT pull=UP
GPIO 47: level=0 fsel=0 func=INPUT pull=UP
GPIO 48: level=0 fsel=0 func=INPUT pull=DOWN
GPIO 49: level=0 fsel=0 func=INPUT pull=DOWN
GPIO 50: level=0 fsel=0 func=INPUT pull=DOWN
GPIO 51: level=0 fsel=0 func=INPUT pull=DOWN
GPIO 52: level=0 fsel=0 func=INPUT pull=DOWN
GPIO 53: level=0 fsel=0 func=INPUT pull=DOWN

vcdbg log messages
------------------

vcdbg: error while loading shared libraries: libdebug_sym.so: cannot open shared object file: No such file or directory
```

(see below for dmesg)

### Logs

Running the `kms_writeback` test:

```
% sudo build/tests/kms_writeback
IGT-Version: 1.26-g38f656fd (arm) (Linux: 5.10.63-v7l+ armv7l)
Starting subtest: writeback-pixel-formats
Subtest writeback-pixel-formats: SUCCESS (0.000s)
Starting subtest: writeback-invalid-parameters
Subtest writeback-invalid-parameters: SUCCESS (0.006s)
Starting subtest: writeback-fb-id
Subtest writeback-fb-id: SUCCESS (10.433s)
Starting subtest: writeback-check-output
(kms_writeback:6450) CRITICAL: Test assertion failure function get_and_wait_out_fence, file ../tests/kms_writeback.c:254:
(kms_writeback:6450) CRITICAL: Failed assertion: ret == 0
(kms_writeback:6450) CRITICAL: Last errno: 38, Function not implemented
(kms_writeback:6450) CRITICAL: sync_fence_wait failed: Timer expired
Stack trace:
  #0 ../lib/igt_core.c:1752 __igt_fail_assert()
  #1 ../tests/kms_writeback.c:254 writeback_sequence()
  #2 ../tests/kms_writeback.c:325 __igt_unique____real_main345()
  #3 ../tests/kms_writeback.c:345 main()
  #4 ./csu/libc-start.c:342 __libc_start_main()
  #5 [_start+0x2c]
Subtest writeback-check-output failed.
**** DEBUG ****
(kms_writeback:6450) igt_fb-DEBUG: igt_create_fb_with_bo_size(width=640, height=480, format=XR24(0x34325258), modifier=0x0, size=0)
(kms_writeback:6450) igt_fb-DEBUG: igt_create_fb_with_bo_size(handle=2, pitch=2560)
(kms_writeback:6450) ioctl_wrappers-DEBUG: Test requirement passed: igt_has_fb_modifiers(fd)
(kms_writeback:6450) DEBUG: Test requirement passed: fb_id > 0
(kms_writeback:6450) igt_kms-DEBUG: display: A.0: plane_set_fb(229)
(kms_writeback:6450) igt_kms-DEBUG: display: A.0: plane_set_size (640x480)
(kms_writeback:6450) igt_kms-DEBUG: display: A.0: fb_set_position(0,0)
(kms_writeback:6450) igt_kms-DEBUG: display: A.0: fb_set_size(640x480)
(kms_writeback:6450) igt_kms-DEBUG: display: Writeback-1: output_set_writeback_fb(226)
(kms_writeback:6450) igt_kms-DEBUG: display: commit {
(kms_writeback:6450) igt_kms-DEBUG: display:     populating plane data: A.0, fb 229
(kms_writeback:6450) igt_kms-DEBUG: plane A.0: Setting property "SRC_X" to 0x0/0
(kms_writeback:6450) igt_kms-DEBUG: plane A.0: Setting property "SRC_Y" to 0x0/0
(kms_writeback:6450) igt_kms-DEBUG: plane A.0: Setting property "SRC_W" to 0x2800000/41943040
(kms_writeback:6450) igt_kms-DEBUG: plane A.0: Setting property "SRC_H" to 0x1e00000/31457280
(kms_writeback:6450) igt_kms-DEBUG: plane A.0: Setting property "CRTC_W" to 0x280/640
(kms_writeback:6450) igt_kms-DEBUG: plane A.0: Setting property "CRTC_H" to 0x1e0/480
(kms_writeback:6450) igt_kms-DEBUG: plane A.0: Setting property "FB_ID" to 0xe5/229
(kms_writeback:6450) igt_kms-DEBUG: plane A.0: Setting property "CRTC_ID" to 0x37/55
(kms_writeback:6450) igt_kms-DEBUG: plane A.0: Setting property "COLOR_ENCODING" to 0x1/1
(kms_writeback:6450) igt_kms-DEBUG: plane A.0: Setting property "COLOR_RANGE" to 0x0/0
(kms_writeback:6450) igt_kms-DEBUG: display:     Writeback-1: preparing atomic, pipe: A
(kms_writeback:6450) igt_kms-DEBUG: Writeback-1: Setting property "WRITEBACK_FB_ID" to 0xe2/226
(kms_writeback:6450) igt_kms-DEBUG: Writeback-1: Setting property "WRITEBACK_OUT_FENCE_PTR" to 0x1aaa02c/27959340
(kms_writeback:6450) igt_kms-DEBUG: display: }
(kms_writeback:6450) CRITICAL: Test assertion failure function get_and_wait_out_fence, file ../tests/kms_writeback.c:254:
(kms_writeback:6450) CRITICAL: Failed assertion: ret == 0
(kms_writeback:6450) CRITICAL: Last errno: 38, Function not implemented
(kms_writeback:6450) CRITICAL: sync_fence_wait failed: Timer expired
(kms_writeback:6450) igt_core-INFO: Stack trace:
(kms_writeback:6450) igt_core-INFO:   #0 ../lib/igt_core.c:1752 __igt_fail_assert()
(kms_writeback:6450) igt_core-INFO:   #1 ../tests/kms_writeback.c:254 writeback_sequence()
(kms_writeback:6450) igt_core-INFO:   #2 ../tests/kms_writeback.c:325 __igt_unique____real_main345()
(kms_writeback:6450) igt_core-INFO:   #3 ../tests/kms_writeback.c:345 main()
(kms_writeback:6450) igt_core-INFO:   #4 ./csu/libc-start.c:342 __libc_start_main()
(kms_writeback:6450) igt_core-INFO:   #5 [_start+0x2c]
****  END  ****
Subtest writeback-check-output: FAIL (31.962s)
```

The tail of `dmesg` after running that test:

```
[1487552.228159] Console: switching to colour dummy device 80x30
[1487552.228271] [IGT] kms_writeback: executing
[1487552.366267] [IGT] kms_writeback: starting subtest writeback-pixel-formats
[1487552.366631] [IGT] kms_writeback: starting subtest writeback-invalid-parameters
[1487552.372378] [IGT] kms_writeback: starting subtest writeback-fb-id
[1487562.803235] [drm:drm_atomic_helper_wait_for_flip_done [drm_kms_helper]] *ERROR* [CRTC:55:crtc-0] flip_done timed out
[1487562.805395] [IGT] kms_writeback: starting subtest writeback-check-output
[1487573.043350] [drm:drm_atomic_helper_wait_for_dependencies [drm_kms_helper]] *ERROR* [CRTC:55:crtc-0] flip_done timed out
[1487583.283472] [drm:drm_atomic_helper_wait_for_dependencies [drm_kms_helper]] *ERROR* [CONNECTOR:48:Writeback-1] flip_done timed out
[1487593.523627] [drm:drm_atomic_helper_wait_for_flip_done [drm_kms_helper]] *ERROR* [CRTC:55:crtc-0] flip_done timed out
[1487605.043737] [drm:drm_atomic_helper_wait_for_dependencies [drm_kms_helper]] *ERROR* [CRTC:55:crtc-0] flip_done timed out
[1487615.283846] [drm:drm_atomic_helper_wait_for_dependencies [drm_kms_helper]] *ERROR* [PLANE:49:plane-0] flip_done timed out
[1487615.286031] [IGT] kms_writeback: exiting, ret=98
[1487615.287706] Console: switching to colour frame buffer device 240x67
[1487625.523962] [drm:drm_atomic_helper_wait_for_dependencies [drm_kms_helper]] *ERROR* [CONNECTOR:48:Writeback-1] flip_done timed out
```

(I can include the entire `dmesg` if desired, perhaps after a clean reboot, if that's helpful.)

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by mripard**
Hi,

Here's a few fixes to fix the timeouts for the IGT writeback tests.

The last patch is probably going to be controversial, and I'm not sure what the right solution could be.
I've started a thread on the mailing lists to see if it's appropriate of if we should find a better solution:
https://lore.kernel.org/dri-devel/20220304144607.3pwfwhxf4zsckan7@houat/T/#u

Fixes #4874

### CODE REVIEW

**Comment by popcornmix** [comment]
By padding, you are referring to the X byte in XRGB8888 (rather than unused pixels if stride is wider than width)?
The transposer will write something to that (it's generally inefficient to preserve what was there, requiring a read/modify/write with wide accesses). I guess it's convention whether you say 0 or ff is the right answer.
In theory it's unused so shouldn't matter, but may be more efficient if we make the same choice as most software would like.

**Comment by mripard** [comment]
> By padding, you are referring to the X byte in XRGB8888 (rather than unused pixels if stride is wider than width)?

Yes

> The transposer will write something to that (it's generally inefficient to preserve what was there, requiring a read/modify/write with wide accesses). I guess it's convention whether you say 0 or ff is the right answer. In theory it's unused so shouldn't matter, but may be more efficient if we make the same choice as most software would like.

It's not really clear to me what most software would like :) Similarly, I'm not sure whether IGT expectations that all hardware can force the X component to 0xff is reasonable. We'll see how the discussion turns out

**Comment by 6by9** [comment]
All look reasonable to me.

Minor grammatical niggle on
> The documentation explicitly states we must prevent the output 2 and 3 from feeding from the same HVS channel.

"The documentation explicitly states we must prevent one HVS channel feeding multiple outputs simultaneously.

Let's add a warning to make some noise if we ever find ourselves in such a case on output 2"

Or we could add a check on each of the channels just to make sure.

In an RGBX format you would hope that the X really was ignored, but acknowledge that for things like IGT it is far easier to just hash the entire buffer, however that does then put expectations on how the writeback connector should function.

**Comment by popcornmix** [comment]
Consensus here is ff for the unused X/alpha value is as good as any value.
If you were to use this buffer as a format with alpha, then ff at least means it would be visible rather than fully transparent (e.g. copying it to /dev/fb configured as ARGB8888).

File: drivers/gpu/drm/vc4/vc4_kms.c [hunk 1]
```diff
@@ -827,16 +830,29 @@ static int vc4_pv_muxing_atomic_check(struct drm_device *dev,
...
+		}
 
 		/* Muxing will need to be modified, mark it as such */
 		new_vc4_crtc_state->update_muxing = true;
 
 		/* If we're disabling our CRTC, we put back our channel */
 		if (!new_crtc_state->enable) {
 			channel = old_vc4_crtc_state->assigned_channel;
+
+			drm_dbg(dev, "%s: Disabling, Freeng channel %d\n",
```
Comment (PeterPablo) [inline drivers/gpu/drm/vc4/vc4_kms.c]: tiny typo: I assume "Freeing" instead of "Freeng" was intended.

**Pull request merged**